src/cov/cov-exclude.c

```diff
@@ -31,102 +31,344 @@ struct _cover_exclude_ctx {
+                   "    src->data:           %d\n"
+                   "    tgt->data:           %d\n",
+                   tgt_prefix_len,
+                   src_prefix_len,
+                   istr(tgt->hier),
+                   istr(src->hier),
+                   istr(tgt_suffix_hier),
+                   istr(src_suffix_hier),
+                   src->data,
+                   tgt->data);
+#endif
+
+            assert(tgt->kind == src->kind);
+            cover_merge_one_item(tgt, src->data);
+
+#ifdef COVER_DEBUG_FOLD
+            printf("    tgt->data(after):    %d\n\n", tgt->data);
+#endif
+            break;
+         }
+      }
+   }
+
+   // Process sub-scopes
+   for (int i = 0; i < src_scope->children.count; i++) {
+      cover_scope_t *src = src_scope->children.items[i];
+
+      for (int j = 0; j < tgt_scope->children.count; j++) {
+         cover_scope_t *tgt = tgt_scope->children.items[j];
+
+         // Compare hierarchical paths, but strip "tgt_scope" prefix from "tgt",
+         // and "src_scope" from "src". Only suffix of hierarchy needs to be the same!
+         ident_t tgt_suffix_hier = NULL;
+         ident_t src_suffix_hier = NULL;
+
+         if (ident_len(tgt->hier) > tgt_prefix_len)
+            tgt_suffix_hier = ident_new(istr(tgt->hier) + tgt_prefix_len);
+         if (ident_len(src->hier) > src_prefix_len)
+            src_suffix_hier = ident_new(istr(src->hier) + src_prefix_len);
+
+         if (tgt_suffix_hier == src_suffix_hier) {
+
+#ifdef COVER_DEBUG_FOLD
+            printf("Folding coverage scope:\n"
+                   "    tgt_prefix_len:      %d\n"
+                   "    src_prefix_len:      %d\n"
+                   "    tgt->hier:           %s\n"
+                   "    src->hier:           %s\n"
+                   "    tgt_suffix_hier:     %s\n"
+                   "    src_suffix_hier:     %s\n\n",
+                   tgt_prefix_len,
+                   src_prefix_len,
+                   istr(tgt->hier),
+                   istr(src->hier),
+                   istr(tgt_suffix_hier),
+                   istr(src_suffix_hier));
+#endif
+
+            cover_fold_scopes(tgt, src);
+            break;
+         }
+      }
+   }
+}
+
+static void cover_iterate_fold_source(cover_data_t *data, cover_scope_t *tgt_scope,
+                                      cover_scope_t *src_scope, cover_fold_cmd_t *cmd)
+{
+   for (int i = 0; i < src_scope->children.count; i++) {
+      cover_scope_t *curr_scp = src_scope->children.items[i];
+
+#ifdef COVER_DEBUG_FOLD
+      printf("Browsing fold source scope: \n"
+             "    Current:  %s\n"
+             "    Expected: %s\n\n",
+             istr(curr_scp->hier), istr(cmd->source));
+#endif
+
+      if (curr_scp->hier == cmd->source) {
+         cmd->found_source = true;
+         notef("Folding coverage scopes:");
+         notef("        Target - %s", istr(tgt_scope->hier));
+         notef("        Source - %s", istr(curr_scp->hier));
```

Comment:
I don't think it's necessary to be this verbose. Alternatively only enable it when `DEBUG` is defined:

```suggestion
#ifdef DEBUG
         diag_t *d = diag_new(NULL, DIAG_DEBUG);
         diag_printf(d, "folding coverage scopes:");
         diag_hint(d, NULL, "        Target - %s", istr(tgt_scope->hier));
         diag_hint(d, NULL, "        Source - %s", istr(curr_scp->hier));
         diag_emit(d);
#endif
```